src/main/c/src/slf4j.c

```diff
@@ -0,0 +1,184 @@
+|   regardless of the license terms of these independent modules, and to copy
+|   and distribute the resulting combined work under terms of your choice,
+|   provided that every copy of the combined work is accompanied by a complete
+|   copy of the source code of RXTX (the version of RXTX used to produce the
+|   combined work), being distributed under the terms of the GNU Lesser General
+|   Public License plus this exception.  An independent module is a
+|   module which is not derived from or based on RXTX.
+|
+|   Note that people who make modified versions of RXTX are not obligated
+|   to grant this special exception for their modified versions; it is
+|   their choice whether to do so.  The GNU Lesser General Public License
+|   gives permission to release a modified version without this exception; this
+|   exception also makes it possible to release a modified version which
+|   carries forward this exception.
+|
+|   You should have received a copy of the GNU Lesser General Public
+|   License along with this library; if not, write to the Free
+|   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+|   All trademarks belong to their respective owners.
+--------------------------------------------------------------------------*/
+
+#include <ctype.h>
+#include <stdlib.h>
+#include <string.h>
+#include "slf4j.h"
+
+/** \brief The maximum length of a log message. */
+#define MESSAGE_MAX 4096
+
+struct Slf4jContext
+{
+	JNIEnv *env;
+	jobject log;
+	/* We want to treat logging contexts like a stack. A JNI function which
+	 * calls back into Java which calls back into JNI will (if it properly
+	 * calls slf4j__setup_instance()/slf4j__setup_static()) initialize a new
+	 * logging context; we don't want to lose the previous one when we pop back
+	 * up the call stack into the original JNI context. */
+	struct Slf4jContext *prev;
+};
+
+static __thread Slf4jContext *thread_context = NULL;
+
+void slf4j_log(LogLevel level, const char *msg)
+{
+	if (thread_context == NULL)
+	{
+		/* No context? No logging. */
+		return;
+	}
+
+	const char *methodName;
+	switch (level)
+	{
+	case LOG_ERROR:
+		methodName = "error";
+		break;
+	case LOG_WARN:
+		methodName = "warn";
+		break;
+	case LOG_INFO:
+		methodName = "info";
+		break;
+	case LOG_DEBUG:
+		methodName = "debug";
+		break;
+	case LOG_TRACE:
+	default:
+		methodName = "trace";
+		break;
+	}
+
+	/* The report() functions used to just conditionally printf whatever was
+	 * handed to them, so all of those invocations pass strings with trailing
+	 * newlines. We'll trim any trailing whitespace before passing to SLF4J. */
+	int messageLength = strlen(msg);
+	while (isspace(msg[messageLength - 1]))
+	{
+		--messageLength;
+	}
+
+	/* A log message 4K in length is likely already an error. Anything longer
+	 * is _definitely_ an error. */
+	if (messageLength > MESSAGE_MAX)
+	{
+		messageLength = MESSAGE_MAX;
+	}
+
+	char *trimmedMessage = strndup(msg, messageLength);
```

Comment:
`strndup()` is not available when compiling for Windows.

Reply:
Ah, yes – I probably never tried that. Not a big deal to fix.